tensorflow/g3doc/hardware/adding_support/index.md

```diff
@@ -0,0 +1,121 @@
+      * [class ThreadPoolDevice: public LocalDevice](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/common_runtime/threadpool_device.h)
+    * [class RemoteDevice: public Device](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/distributed_runtime/remote_device.cc)
+
+
+There is also a structure referred to as a Device Factory,
+* [class DeviceFactory](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/common_runtime/device_factory.h)
+  * [class ThreadPoolDeviceFactory: public Device Factory](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/common_runtime/threadpool_device_factory.cc)
+  * [class BaseGPUDeviceFactory : public DeviceFactory](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/common_runtime/gpu/gpu_device.h)
+    * [class GPUDeviceFactory: public BaseGPUDeviceFactory](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/common_runtime/gpu/gpu_device_factory.cc)
+  * [class GPUCompatibleCPUDeviceFactory: public DeviceFactory](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/common_runtime/gpu/gpu_device_factory.cc)
+
+
+It is clear that TensorFlow is currently built to support 3 major 
+types of devices: CPU, GPU, Distributed. We will walk throught the 
+creation of a new kind of device for each of these classes and leave 
+it to the reader to decide which general type of device his/her 
+hardware falls into best.
+
+
+## New ThreadPool
+
+We would like to create a new hardware interface that we can use in
+standard TensorFlow operations. For example, we might want a ficticious
+device called an SPU that I can use like,
+
+```python
+import tensorflow as tf
+
+# Creates a graph.                                                              
+with tf.device('/spu:0'):
+  a = tf.constant([1.0, 2.0, 3.0, 4.0, 5.0, 6.0], shape=[2, 3], name='a')
+  b = tf.constant([1.0, 2.0, 3.0, 4.0, 5.0, 6.0], shape=[3, 2], name='b')
+  c = tf.matmul(a, b)
+# Creates a session with log_device_placement set to True.                      
+sess = tf.Session(config=tf.ConfigProto(log_device_placement=True))
+# Runs the op.                                                                  
+print sess.run(c)
+```
+
+with output,
+```bash
+$ python test.py 
+Device mapping: no known devices.
+I tensorflow/core/common_runtime/direct_session.cc:175] Device mapping:
+
+MatMul: /job:localhost/replica:0/task:0/spu:0
+I tensorflow/core/common_runtime/simple_placer.cc:818] MatMul: /job:localhost/replica:0/task:0/spu:0
+b: /job:localhost/replica:0/task:0/spu:0
+I tensorflow/core/common_runtime/simple_placer.cc:818] b: /job:localhost/replica:0/task:0/spu:0
+a: /job:localhost/replica:0/task:0/spu:0
+I tensorflow/core/common_runtime/simple_placer.cc:818] a: /job:localhost/replica:0/task:0/spu:0
+[[ 22.  28.]
+ [ 49.  64.]]
+```
+
+To do so, we need to implement a few classes and modify the C++
```

Comment:
"Adding a new devices requires implementing several classes and interfaces in C++.  Step-by-step instruction can be found [here](fake_device.md), but the overview of the changes is as follows:"